### PR TITLE
[NG] Focus trap directive

### LINKED ISSUE #767: Tabbing within modals should keep the focus on the modal
**Select one ...**  (check one with "x")
```
[X] bug
[] feature request
[ ] enhancement
```

### Expected behavior
When tabbed in the modal, the tab focus should stay within the modal. This helps in better user accessibility.

### Actual behavior
When tabbed in the modal, the tab focus leaves the modal.

### Reproduction of behavior
This behavior can be reproduced in the clarity design website
https://vmware.github.io/clarity/documentation/modals

### Environment details
* **Angular version:** 2.0.X
N/A
* **Clarity version:** 
latest clarity version
* **OS and version:** 
MAC os
* **Browser:** 
Chrome

### PR DESCRIPTION
**Pull request by victormejia**
New `clrFocusTrap` directive to trap the user tabbing to the focusable elements within the container.

Pardon the delay on this one. Created a new PR since it had been a while and things have changed quite a bit. 

@youdz Went ahead and used the `focusin` event. Seems to be working great, however, I did need to add the `tabindex` attribute to the host element in order for it to be focusable. What this implementation also means is that the shift+tab keypress will only cycle up to the host element, and not have a circular rotation. Feedback appreciated!

focustrap

Resolves #767

Signed-off-by: Victor Mejia <victor.a.mejia@gmail.com>

### CODE REVIEW

File: src/clarity-angular/utils/focus-trap/focus-trap.directive.spec.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import {By} from "@angular/platform-browser/";
+
+describe("InputTrap", () => {
+    let fixture: ComponentFixture<any>;
+    let compiled: any;
+    let directive: FocusTrapDirective;
+
+    beforeEach(() => {
+        TestBed.configureTestingModule({
+            imports: [ClarityModule.forRoot()],
```
Comment (hippee-lee) [inline src/clarity-angular/utils/focus-trap/focus-trap.directive.spec.ts]: I believe we jsut need the `ClrModalModule` here.

Comment (youdz) [inline src/clarity-angular/utils/focus-trap/focus-trap.directive.spec.ts]: Actually, you only want to import `ClrFocusTrapModule`. Declaring all of Clarity is overkill and will _severely_ slow down tests.

**Comment by victormejia** [comment]
Thanks for the feedback! Went ahead and updated with requested changes.

File: src/app/modal/modal.demo.routing.ts [hunk 2]
```diff
@@ -34,6 +35,7 @@ const ROUTES: Routes = [
             { path: "backdrop", component: ModalBackdropDemo },
             { path: "sizes", component: ModalSizesDemo },
             { path: "max-height", component: ModalMaxHeightDemo }
```
Comment (mathisscott) [inline src/app/modal/modal.demo.routing.ts]: @victormejia 
TSLint is complaining about a missing comma at the end of line 37 here.

File: src/clarity-angular/utils/focus-trap/index.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+* Copyright (c) 2017 VMware, Inc. All Rights Reserved.
+* This software is released under MIT license.
+* The full license information can be found in LICENSE in the root directory of this project.
+*/
+ import { Type } from "@angular/core";
+ import {FocusTrapDirective} from "./focus-trap.directive";
+
+ export const FOCUS_TRAP_DIRECTIVES: Type<any>[] = [
+     FocusTrapDirective
+ ];
```
Comment (mathisscott) [inline src/clarity-angular/utils/focus-trap/index.ts]: It looks like we have one gemini test failing in CSS regression. I don't think that's a big deal. Maybe @hippee-lee can help get that resolved?

**Comment by victormejia** [comment]
A little unsure what to do with the CSS regression test.

**Comment by mathisscott** [comment]
@victormejia 
@hippee-lee said he would help you get that resolved in our planning meeting yesterday. So if you can take care of the missing comma, then you can wait for him to get the file(s) to you that need to be replaced.

**Comment by hippee-lee** [comment]
Hi @victormejia - thanks for working on this. Attached is a zip archive with an updated reference image for the modal test `max-height`. Please replace the `max-height` folder in your PR branch.

The gemini reference images for modal are located here:
``` YOUR_CLARITY_ROOT/gemini/screens/modal```
All you should need to do is replace the max-height folder with the one attached here.
[max-height.zip](https://github.com/vmware/clarity/files/1131687/max-height.zip)

Here is a screen shot that may help: 
screen shot 2017-07-07 at 9 39 44 am

When you commit  that change and push to your PR branch gemini will re-run and use the updated reference image. Please reach out to me if you have any questions or issues.

**Comment by victormejia** [comment]
thank so much @hippee-lee ! Looks like all is set.

**Pull request merged**